### PR TITLE
Database factory injection

### PR DESCRIPTION
**Pull request by alainbodiguel**

### CODE REVIEW

File: arlas-core/src/main/java/io/arlas/server/impl/elastic/utils/ElasticClient.java [hunk 1]
```diff
@@ -264,7 +263,7 @@ public void close() {
                 this.sniffer.close();
             }
             this.client.close();
```
Comment (laurent-thiebaud-gisaia) [inline arlas-core/src/main/java/io/arlas/server/impl/elastic/utils/ElasticClient.java:265]: Don't see the code around, but shouldn't the `close()`be in a finally?

File: arlas-server/src/main/java/io/arlas/server/app/DatabaseToolsFactory.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+import com.codahale.metrics.health.HealthCheck;
+import io.arlas.server.dao.CollectionReferenceDao;
+import io.arlas.server.ogc.common.dao.OGCCollectionReferenceDao;
+import io.arlas.server.ogc.wfs.services.WFSToolService;
+import io.arlas.server.services.ExploreService;
+
+import java.util.Map;
+
+public abstract class DatabaseToolsFactory {
```
Comment (laurent-thiebaud-gisaia) [inline arlas-server/src/main/java/io/arlas/server/app/DatabaseToolsFactory.java:30]: Does the name match what it does? It looks to be more than tools

File: arlas-server/src/main/java/io/arlas/server/impl/ElasticDatabaseToolsFactoryImpl.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    private final CollectionReferenceDao collectionReferenceDao;
+    private final OGCCollectionReferenceDao ogcDao;
+    private final WFSToolService wfsService;
+
+    public ElasticDatabaseToolsFactoryImpl(ArlasServerConfiguration configuration) {
+        super(configuration);
+        this.elasticClient = new ElasticClient(configuration.elasticConfiguration);
+        this.exploreService = new ElasticExploreServiceImpl(elasticClient, configuration);
+        this.collectionReferenceDao = new ElasticCollectionReferenceDaoImpl(elasticClient, configuration.arlasindex, configuration.arlascachesize, configuration.arlascachetimeout);
+        this.ogcDao = new ElasticOGCCollectionReferenceDaoImp(elasticClient, collectionReferenceDao, configuration.arlasindex, Service.CSW);
```
Comment (laurent-thiebaud-gisaia) [inline arlas-server/src/main/java/io/arlas/server/impl/ElasticDatabaseToolsFactoryImpl.java:52]: ElasticOGCCollectionReferenceDaoImp => ElasticOGCCollectionReferenceDaoImpl

File: arlas-server/src/main/java/io/arlas/server/app/DatabaseToolsFactory.java (hunk 2 above)
Comment (alainbodiguel) [reply arlas-server/src/main/java/io/arlas/server/app/DatabaseToolsFactory.java:30]: Any other idea would be welcome!

File: arlas-server/src/main/java/io/arlas/server/impl/ElasticDatabaseToolsFactoryImpl.java (hunk 3 above)
Comment (alainbodiguel) [reply arlas-server/src/main/java/io/arlas/server/impl/ElasticDatabaseToolsFactoryImpl.java:52]: ok

File: arlas-core/src/main/java/io/arlas/server/impl/elastic/utils/ElasticClient.java (hunk 1 above)
Comment (alainbodiguel) [reply arlas-core/src/main/java/io/arlas/server/impl/elastic/utils/ElasticClient.java:265]: sniffer.close does not raise any exception, while client.close() raises an IOException which is caught.

File: arlas-core/src/main/java/io/arlas/server/impl/elastic/utils/ElasticClient.java [hunk 4]
```diff
@@ -78,12 +79,13 @@ public RestHighLevelClient getClient() {
         return client;
     }
 
-    public ClusterHealthResponse health() throws ArlasException {
+    public boolean isClusterHealthRed() throws ArlasException {
         try {
-            return client.cluster().health(new ClusterHealthRequest(), RequestOptions.DEFAULT);
+            ClusterHealthResponse response = client.cluster().health(new ClusterHealthRequest(), RequestOptions.DEFAULT);
+            return response.getStatus() != ClusterHealthStatus.RED;
```
[the remark this answers is not in the document]
Comment (alainbodiguel) [reply arlas-core/src/main/java/io/arlas/server/impl/elastic/utils/ElasticClient.java:85]: well spotted!

File: arlas-core/src/main/java/io/arlas/server/impl/elastic/utils/ElasticClient.java [hunk 5]
```diff
@@ -43,18 +41,21 @@
...
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import java.io.IOException;
+import java.util.Arrays;
 import java.util.HashMap;
 import java.util.LinkedHashMap;
 import java.util.Map;
 
+@SuppressWarnings({"rawtypes"})
```
[the remark this answers is not in the document]
Comment (alainbodiguel) [reply arlas-core/src/main/java/io/arlas/server/impl/elastic/utils/ElasticClient.java:56]: yes, there are several methods concerned in this class

File: arlas-core/src/main/java/io/arlas/server/impl/elastic/dao/ElasticCollectionReferenceDaoImpl.java [hunk 6]
```diff
@@ -205,8 +217,7 @@ public CollectionReference putCollectionReference(CollectionReference collection
         }
     }
 
-
-
+    @SuppressWarnings({"unchecked", "rawtypes"})
```
[the remark this answers is not in the document]
Comment (alainbodiguel) [reply arlas-core/src/main/java/io/arlas/server/impl/elastic/dao/ElasticCollectionReferenceDaoImpl.java:220]: yes, there are several methods concerned in this class

File: arlas-core/src/main/java/io/arlas/server/app/ArlasServerConfiguration.java [hunk 7]
```diff
@@ -243,11 +186,9 @@ public void check() throws ArlasConfigurationException {
             arlasAuthConfiguration = new ArlasAuthConfiguration();
             arlasAuthConfiguration.enabled = false;
         }
-        if (elasticEnableSsl == null) {
-            elasticEnableSsl = false;
-        }
-        if (elasticCompress == null) {
-            elasticCompress = true;
+
+        if (arlasDatabaseFactoryClass == null) {
```
[the remark this answers is not in the document]
Comment (alainbodiguel) [reply arlas-core/src/main/java/io/arlas/server/app/ArlasServerConfiguration.java:190]: I'm not sure if the required = true of the JsonProperty does that when I read the documentation.
So I've done the as for other properties above in this class.

File: arlas-server/src/main/java/io/arlas/server/app/DatabaseToolsFactory.java (hunk 2 above)
Comment (laurent-thiebaud-gisaia) [reply arlas-server/src/main/java/io/arlas/server/app/DatabaseToolsFactory.java:30]: I guess it should be something like DatabaseServicesFactory, and dao should only be visible from a service-related class. Or we could use dependency injection. But as it is not like this today, I have no better idea neither.

Comment (alainbodiguel) [reply arlas-server/src/main/java/io/arlas/server/app/DatabaseToolsFactory.java:30]: Yes, I'd like to go further into the refactoring, but I'm going slowly in order to limitate the risk and ease the PRs. We'll change the name later then.

**Comment by MohamedHamouGisaia** [comment]
LGTM !
Only an entry `arlas_database_factory_class` is missing in `docs/arlas-server-configuration.md` file

**Pull request merged**